runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -84,11 +84,11 @@ protected ILogData removeFromQueue(NavigableSet<Long> queue) {
                 // to be serviced immediately, rather than reading one entry at a time.
                 ld = read(currentRead, queue);
 
-                if (queue == getCurrentContext().readQueue && ld != null) {
+                if (queue == getCurrentContext().readQueue) {
                     // Validate that the data entry belongs to this stream, otherwise, skip.
                     // This verification protects from sequencer regression (tokens assigned in an older epoch
                     // that were never written to, and reassigned on a newer epoch)
-                    if (ld.containsStream(this.id) && ld.getType() == DataType.DATA) {
+                    if (ld.isHole() || (ld.containsStream(this.id) && ld.isData())) {
```

Comment:
I am a bit confused, why do we want to surface the hole itself to the object layer? wouldn't that give the wrong behavior? If we are enforcing holes for different reasons (materialize a stream or any other reason), a client should not perceive it...If we are directly consuming the stream it would perceive them right?

Reply:
Do you mean we should add hole to ResolvedQueue in AbstractQueuedStreamView?

Reply:
We are adding it here, right? Which I believe is correct... My question is more like, when it's a hole readNext is being set to false meaning that the upper layer receives this hole. So if you are directly manipulating a stream a nextUpTo would return a hole at some point, or is it further filtered? Maybe I did not see when we filter it in upper layers

Reply:
@zhangn49 This is what I mean (run this test). To someone consuming the stream directly we are exposing holes (null entries) which should be transparent to a reader (these holes are our way to avoid other issues with checkpoints):


    @Test
    public void testNextUpTo() {
        final int numWrites = 10;

        byte[] testPayload = "hello world".getBytes();

        String stream = "stream1";
        UUID id1 = CorfuRuntime.getStreamID(stream);

        StreamingMap<String, String> map = r.getObjectsView()
                .build()
                .setStreamName(stream)
                .setTypeToken(new TypeToken<CorfuTable<String, String>>() {})
                .open();

        CorfuRuntime producer = getNewRuntime(getDefaultNode())
                .connect();

        CorfuRuntime consumer = getNewRuntime(getDefaultNode())
                .connect();

        IStreamView svProducer = producer.getStreamsView().get(id1);
        IStreamView svConsumer = consumer.getStreamsView().get(id1);


        for (int x = 0; x < numWrites; x++) {
            svProducer.append(testPayload);
        }

        // Emulate the extra Token the Checkpointer requests before appending a checkpoint
        // producer.getSequencerView().next(id1);
        // producer.getAddressSpaceView().write(new Token(0, numWrites), LogData.getHole(numWrites));

        // Get two entries before checkpointing (so we don't load from checkpoint on next access)
        assertThat(svConsumer.nextUpTo(numWrites).getPayload(consumer)).isEqualTo(testPayload);
        assertThat(svConsumer.nextUpTo(numWrites).getPayload(consumer)).isEqualTo(testPayload);

        MultiCheckpointWriter checkpointWriter = new MultiCheckpointWriter();
        checkpointWriter.addMap(map);
        checkpointWriter.appendCheckpoints(r, "Author");

        // Consume Until Tail
        long tail = consumer.getSequencerView().query(id1);
        System.out.println("Tail is: " + tail);

        ILogData data;
        while (svConsumer.hasNext()) {
            data = svConsumer.nextUpTo(tail);
            System.out.println("Next @" + data.getGlobalAddress() + " bp is: " + data.getBackpointer(id1));
            assertThat(data.getPayload(consumer)).isEqualTo(testPayload);

        }

    }

Reply:
I see. Since we need VLO's version to point to a hole to avoid unnecessary sync, we are adding hole to streamView. 

getNextEntries is safe now, since we filter hole out.
For nextUpTo/next, it will return a hole to direct sv consumers. It seems we don't have direct consumer right now.

I am thinking if I can do this in nextUpTp, to skip hole.
```
        if (entry != null) {
            // Update the pointer.
            updatePointer(entry);

            // Check if entry is a hole
            if (entry.isHole()) {
                return nextUpTo(maxGlobal);
            }

            // Process the next entry, checking if the context has changed.
            // If the context has changed, we read again, since this entry
            // does not contain any data, and we need to follow the new
            // context.
            if (processEntryForContext(entry)) {
                return nextUpTo(maxGlobal);
            }
        }
```

In this case, hasNext() will return true when only hole left in this stream, but looks still consistent with its definition.

```
/** Returns whether or not there are potentially more entries in this
     * stream - this function may return true even if there are no entries
     * remaining, as addresses may have been acquired by other clients
     * but not written yet, or the addresses were hole-filled, or just failed.
     * @return      True, if there are potentially more entries in the stream.
     */
    boolean hasNext();
```

Reply:
I Think all you need to do, is in the AddressMapStreamView removeFromQueue, if it is a hole, set readNext=True... and this will automatically filter the hole for you to upper layers. So do something like:


if (ld.isHole() || (ld.containsStream(this.id) && ld.isData())) {
                        addToResolvedQueue(getCurrentContext(), currentRead);
                        readNext = ld.isHole()? true: false;		                         
}

Reply:
Discussed offline, skip hole in nextUpTo to update pointer.
Added comment why we add hole and skip it, also added this unit test to StreamViewTest.